### PR TITLE
Supports SSM Parameters as serverless variable references

### PR DESCRIPTION
**Pull request by lorengordon**
## What did you implement:

Adds ability to specify SSM parameters as serverless variable references.

This is mostly intended to support discussion in #4053, so everyone can see what the code looks like.

## How did you implement it:

* Created a new regex to match `ssm:` variable references
* Created a new function that calls the SSM API to retrieve the parameter value

I cribbed a lot of the implementation from the already existing code for `s3:` variable references.

In order to support decryption of SecureString parameters, I selected the `~` as a delimiter between the parameter name and a true/false value that maps to the `WithDecryption` option of the SSM GetParameter API. I selected the `~` only because it is already allowed in the default variable regex, and because it is not allowed as a character in the SSM parameter name.

The AWS SDK needed to be updated to version 2.75.0 because that is when the GetParameter API was [added](https://github.com/aws/aws-sdk-js/blob/master/CHANGELOG.md#2750).

## How can we verify it:

Here are commands to put a parameter into SSM:
```
aws ssm put-parameter --name foo --value bar --type SecureString
aws ssm put-parameter --name bar --value foo --type String
aws ssm put-parameter --name /foo/bar --value bar --type SecureString
aws ssm put-parameter --name /bar/foo --value foo --type String
```

With a minimal serverless.yml file, you would then be able to reference the variables like so:

```
${ssm:foo}
${ssm:bar~true}
${ssm:/foo/bar}
${ssm:/bar/foo~true}
```

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

**Comment by lorengordon** [comment]
If the discussion settles on this approach, I'll of course write tests and docs. :D

**Comment by pmuens** [comment]
Thanks again for PRing this @lorengordon 👍

> If the discussion settles on this approach, I'll of course write tests and docs. :D

That sounds like a good plan 💪!

**Comment by lorengordon** [comment]
I had some time this morning, and seems like the discussion hasn't yet identified a method of incorporating this as a plugin, so I went ahead and added the tests and docs.

**Comment by lorengordon** [comment]
I'll correct the lint failures...

**Comment by eahefnawy** [comment]
Following the discussion in the issue, we'll close this PR as it's now longer needed.

Thanks again @lorengordon 😊

**Comment by vladgolubev** [comment]
Looks like you spent more time speculating in that discussion than you would ever spend maintaining this 23-lines-of-code feature. I am frankly disappointed.

**Comment by eahefnawy** [comment]
@vladgolubev discussions are always welcome 😊 

Just for clarification though, based my experience maintaining this repo for two years now, it's never about the LOC. Few LOC could cause serious hours of debugging in the long term. It's never just about reviewing and testing the PR, but keeping it working and up to date with every single change we make from now until the project dies. We've learned to be careful of any unnecessary bits of code.

If the community show a real need for this use case, it'd def be worth maintaining and reopening! 💪 ... but for now, it'd be very easy to just copy this script and use the generic "js files" reference.

Sorry for the discouragement. Hope you guys understand 🙏

**Comment by alexdebrie** [comment]
@lorengordon Thanks for doing the leg work to build out this feature. I was very excited to see this feature implemented when I saw the PR opened. I should have communicated that earlier to my Serverless teammates 😄 .

I see "how do I manage secrets with Serverless" as a consistent issue, and something that I haven't entirely settled on. I've used the [serverless-kms-plugin](https://github.com/SC5/serverless-kms-secrets) which works fine, but I'd prefer something more lightweight that is built directly into the Framework. Plus, it's nice to store secrets in a central place like SSM rather than in encrypted values scattered across different repositories.

IMO, part of the problem here is that SSM is still relatively unknown. It's pretty hidden in the AWS docs and some of the details aren't fleshed out (such as, does it cost anything?? My guess is not, but what's stopping me from using it as a free key-value service for any sort of data? 😄 ).

That being said, I'm seeing more and more people talk about SSM. In fact, _just today_, Segment posted a blog post featuring SSM titled [The Right Way to Manage Secrets in AWS](https://segment.com/blog/the-right-way-to-manage-secrets/). I'm leaning more toward using it widely, and it'd be nice for this to be a first-class feature in the Framework.

I understand @eahefnawy and @pmuens concerns about maintenance, but IMHO this is a something that should be included. I'll be making my case. If it doesn't get added, I'd be happy to assist on a plugin with you.

**Comment by eahefnawy** [comment]
@lorengordon we have discussed this internally and we concluded that this could be useful for many users and would answer the secret management questions we get a lot. Initially I didn't think about it this way, apologies for that 😞 , but thanks to @alexdebrie thoughts (which are backed with data! 😄), I think it'd be better to reopen this PR.

@lorengordon hope you understand our process and concerns by now and hope there are no hard feelings 😊 ... and thanks **A LOT** for the effort you put into this ☺️

**Comment by lorengordon** [comment]
@eahefnawy no hard feelings! I apologize if I came off that way. I was probably hungry. 😆 

Looking back, I probably should have mentioned secrets management more explicitly, rather than just external management of variables/parameters in general.

I'll rebase from an updated master branch shortly...

**Comment by lorengordon** [comment]
Far as the code goes, like I said I copied the approach from `getValueFromS3()`. However, I'm seeing that it is incompatible with the variable override syntax, since it [throws an exception](https://github.com/serverless/serverless/pull/4062/files#diff-b921cbcedd2151d24391fb39acb84adeR373) if the request returns an error. Could someone please demonstrate the correct way to do this? 

Here is an example to test, just using the default "hello" template:

```
service: hello

provider:
  name: aws
  runtime: nodejs6.10

functions:
  hello:
    handler: handler.hello
    events:
      - s3:
          bucket: ${ssm:/path/to/bucket, "foo"}
          event: s3:ObjectCreated:Put
```

File: package.json [hunk 1]
```diff
@@ -91,7 +91,7 @@
     "apollo-client": "^1.4.2",
     "archiver": "^1.1.0",
     "async": "^1.5.2",
-    "aws-sdk": "^2.7.13",
```
Comment (pmuens) [inline package.json]: Could you update the `package-lock.json` file as well?

Comment (lorengordon) [reply package.json]: Certainly. Is there a proper way to do it? I haven't yet figured out exactly how npm manages that stuff.

**Comment by eahefnawy** [comment]
@lorengordon @pmuens I just tried this with S3 and it's an issue there as well! Really nice catch! 👍 

@lorengordon instead of following S3 code, would be great if you fix this issue here as a starting point. What I think should happen is that if the parameter store doesn't exist (or frankly any other error), instead of throwing an error, just throw a warning. You'll need to update and use the `warnIfNotFound` method. This is regardless you're providing a fallback variable or not.

Incorrect references shouldn't throw an error. And with the help of the print command in #4169 debugging variable issues will be a lot easier!

File: package.json (hunk 1 above)
Comment (pmuens) [reply package.json]: Trust me. I faced this situation a couple of times and I don't know the correct answer to this question (if there's even a correct answer...)

Here's what I do all the time: `rm -rf node_modules package-lock.json && npm install` 😅.

Comment (lorengordon) [reply package.json]: Done.

**Comment by lorengordon** [comment]
@eahefnawy The last commit adds support for the overwrite syntax. Basically checks for the expected error if the SSM parameter does not exist and simply returns `undefined` in that case. Other error messages still throw an exception.

Let me know if this looks acceptable... I have literally no idea what this promise stuff is about, and I'm really just going by trial and error.

**Comment by lorengordon** [comment]
Apparently I really need some guidance on these tests. I don't know how the last commit reduces coverage any; I actually added another test to try to cover the difference between an expected error and an unexpected error. At this point, I think I'm going to need an actual diff to do any better. This javascript syntax is impenetrable.

**Comment by eahefnawy** [comment]
Awesome! I've also tested the overwrite syntax and it's working great. G2M. 😊 

Thanks again @lorengordon 🙌

**Comment by mazzy89** [comment]
@lorengordon I've tested with an use case and it fails. In the case that the parameter is hierarchical and looks like this `/serverless/the-leve/staging/my_param` the `groups[1]` in the func https://github.com/serverless/serverless/pull/4062/commits/a266d4f224ae1ada619613f11685eed5fc4a9dbe#diff-b921cbcedd2151d24391fb39acb84adeR360 is not correct and then the param is not fetched correcty

**Comment by lorengordon** [comment]
Strange, I tested hierarchical parameters explicitly and it worked fine. Seems to still be working now. Not sure what you might be hitting.

serverless.yml:
```
service: hello

provider:
  name: aws
  runtime: nodejs6.10
  environment:
    FOOBAR: ${ssm:/foo/bar}
    BARFOO: ${ssm:/bar/foo}

functions:
  hello:
    handler: handler.hello
```

proof:
```
> aws ssm put-parameter --name /foo/bar --value bar --type String
> aws ssm put-parameter --name /bar/foo --value foo --type String
> serverless package
Serverless: Packaging service...
Serverless: Excluding development dependencies...
> grep FOO .\.serverless\cloudformation-template-update-stack.json
            "FOOBAR": "bar",
            "BARFOO": "foo"
```

**Comment by mazzy89** [comment]
@lorengordon please try with a parameter like this `/serverless/the-level/staging/my_param` and you will understand what I mean.your use case works like a charm. when the param has a dash then it is broken

**Comment by lorengordon** [comment]
I see, it's the hyphen in the param name. Please open a bug report.

**Comment by mazzy89** [comment]
sure. I'll do it

**Comment by pmuens** [comment]
Nice catch @mazzy89 👍

Would be super nice if you could open a new issue which describes the bug in detail!

**Comment by lorengordon** [comment]
Easy fix: https://github.com/serverless/serverless/pull/4249

**Comment by pmuens** [comment]
> Easy fix: #4249

Thanks for working on the fix @lorengordon 👍 💯

**Comment by tafs7** [comment]
@lorengordon @eahefnawy This is an AWESOME feature!  I've been looking for something like this and it fulfills an immediate need I have.  Thank you!

One question - will any of the documentation be updated to reflect (e.g. serverless.yml page).  If it weren't for this Github issue and a recent blog post on Hacker Moon (https://hackernoon.com/you-should-use-ssm-parameter-store-over-lambda-env-variables-5197fc6ea45b), I would not have found out about this feature.  I think it deserves to be highlighted in the docs for the Serverless Framework.

**Comment by lorengordon** [comment]
@tafs7 Ask and your wish shall be granted. Docs were actually included with this PR... :)

https://serverless.com/framework/docs/providers/aws/guide/variables#reference-variables-using-the-ssm-parameter-store

**Pull request merged**